src/fastpath/fastpath.c

```diff
@@ -6,15 +6,254 @@
+        slowpath(SysSend);
+    }
+
+    /* Check that we are allowed to send to this cap */
+    if (unlikely(!cap_notification_cap_get_capNtfnCanSend(cap))) {
+        slowpath(SysSend);
+    }
+
+    /* Get the notification address */
+    notification_t *ntfnPtr = NTFN_PTR(cap_notification_cap_get_capNtfnPtr(cap));
+
+    /* Get the notification state */
+    uint32_t ntfnState = notification_ptr_get_state(ntfnPtr);
+
+    /* Get the notification badge */
+    word_t badge = cap_notification_cap_get_capNtfnBadge(cap);
+    switch (ntfnState) {
+    case NtfnState_Active: {
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+        ksKernelEntry.is_fastpath = true;
+#endif
+        ntfn_set_active(ntfnPtr, badge | notification_ptr_get_ntfnMsgIdentifier(ntfnPtr));
+        restore_user_context();
+        UNREACHABLE();
+    }
+    case NtfnState_Idle: {
+        tcb_t *dest = (tcb_t *) notification_ptr_get_ntfnBoundTCB(ntfnPtr);
+
+        /* Check if we are bound and that thread is waiting for a message */
+        if (dest && thread_state_ptr_get_tsType(&dest->tcbState) == ThreadState_BlockedOnReceive) {
+            /* Checks for maybeDonateSchedContext(). Check whether the thread already has
+             * a SC or if one can be donated from the notification. If neither is true, go to
+             * slowpath */
+            if (!dest->tcbSchedContext) {
+                sc = SC_PTR(notification_ptr_get_ntfnSchedContext(ntfnPtr));
+                if (sc == NULL || sc->scTcb != NULL) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+
+                /* Slowpath the case where dest has its FPU context in the FPU of a core*/
+#ifdef ENABLE_SMP_SUPPORT
+#ifdef CONFIG_HAVE_FPU
+                if (nativeThreadUsingFPU(dest)) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+#endif
+#endif
+            } else {
+                sc = dest->tcbSchedContext;
+            }
+
+            /* Only fastpath signal to threads which will not become the new highest prio thread on the
+             * core of their SC, even if the currently running thread on the core is the idle thread. */
+            if (NODE_STATE_ON_CORE(ksCurThread, sc->scCore)->tcbPriority < dest->tcbPriority) {
+                slowpath(SysSend);
+            }
+
+            /* Simplified schedContext_resume that does not change state and reverts to the
+            * slowpath in cases where the SC does not have sufficient budget, as this case
+            * adds extra scheduler logic. Normally, this is done after donation of SC
+            * but after tweaking it, I don't see anything executed in schedContext_donate
+            * that will affect the conditions of this check */
+            if (sc->scRefillMax > 0) {
+                if (!(refill_ready(sc) && refill_sufficient(sc, 0))) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
```

Comment:
The convention seems to be to have no `UNREACHABLE()` after `slowpath()` calls.

Reply:
You're right. Not sure where I picked that up from.

Reply:
Do you think UNREACHABLE() is appropriate for other points where it exits the kernel, such as line 64?

Reply:
restore_user_context has some extra IPC-specific behavior related to transferring a message to the receiver thread.

Reply:
But your code is doing signals, not IPC, so why is it necessary? And the IPC code is not using it, so why would the signal fastpath need it? A comment explaining that would help, because it is counter-intuitive.

Reply:
Sorry, I edited the comment after posting it, which may have led to some confusion. `fastpath_restore` is the one with IPC-specific behavior, not `restore_user_context`

Reply:
Ah, that makes much more sense. A comment would still help I think, or `fastpath_restore` should be renamed to `fastpath_ipc_restore`.

Reply:
If it's appropriate, I can submit another PR to rename fastpath_restore as it seems a bit tangential to this one.

Reply:
I wouldn't bother, adding a comment before the `restore_user_context` call at the end of `fastpath_signal` is sufficient. or just leave it as it is.